test/e2e-go/upgrades/rekey_support_test.go

```diff
@@ -0,0 +1,121 @@
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/test/framework/fixtures"
+)
+
+// TestRekeyUpgrade tests that we rekey does not work before the upgrade and works well after
+func TestRekeyUpgrade(t *testing.T) {
+	a := require.New(t)
+
+	// set the small lambda to 500 for the duration of this test.
+	roundTimeMs := 500
+	lambda := os.Getenv("ALGOSMALLLAMBDAMSEC")
+	os.Setenv("ALGOSMALLLAMBDAMSEC", fmt.Sprintf("%d", roundTimeMs))
+	defer func() {
+		if lambda == "" {
+			os.Unsetenv("ALGOSMALLLAMBDAMSEC")
+		} else {
+			os.Setenv("ALGOSMALLLAMBDAMSEC", lambda)
+		}
+	}()
+
+	consensus := makeApplicationUpgradeConsensus(t)
+
+	var fixture fixtures.RestClientFixture
+	fixture.SetConsensus(consensus)
+	fixture.Setup(t, filepath.Join("nettemplates", "TwoNodes100SecondTestUnupgradedProtocol.json"))
+	defer fixture.Shutdown()
+
+	client := fixture.GetLibGoalClientForNamedNode("Node")
+	accountList, err := fixture.GetNodeWalletsSortedByBalance(client.DataDir())
+	a.NoError(err)
+
+	accountA := accountList[0].Address
+	wh, err := client.GetUnencryptedWalletHandle()
+	a.NoError(err)
+
+	accountB, err := client.GenerateAddress(wh)
+	a.NoError(err)
+
+	round, err := client.CurrentRound()
+	a.NoError(err)
+
+	// Ensure no rekeying happened
+	ad, err := client.AccountData(accountA)
+	a.NoError(err)
+	a.Equal(basics.Address{}, ad.AuthAddr)
+
+	// rekey A -> B (RekeyTo check)
+	fee := uint64(1000)
+	amount := uint64(1000000)
+	lease := [32]byte{}
+	tx, err := client.ConstructPayment(accountA, accountB, fee, amount, nil, "", lease, basics.Round(round), basics.Round(round+1000))
+	a.NoError(err)
+
+	addrB, err := basics.UnmarshalChecksumAddress(accountB)
+	a.NoError(err)
+	tx.RekeyTo = addrB
+
+	rekey, err := client.SignTransactionWithWalletAndSigner(wh, nil, "", tx)
+	a.NoError(err)
+	_, err = client.BroadcastTransaction(rekey)
+	a.Error(err)
+	require.Contains(t, err.Error(), "transaction has RekeyTo set but rekeying not yet enable")
+
+	// use rekeyed key to authorize (AuthAddr check)
+	tx.RekeyTo = basics.Address{}
+	rekeyed, err := client.SignTransactionWithWalletAndSigner(wh, nil, accountB, tx)
+	a.NoError(err)
+	_, err = client.BroadcastTransaction(rekeyed)
+	a.Error(err)
+	require.Contains(t, err.Error(), "nonempty AuthAddr but rekeying not supported")
+	// go to upgrade
+	curStatus, err := client.Status()
+	require.NoError(t, err)
+	initialStatus := curStatus
+
+	startLoopTime := time.Now()
+
+	// wait until the network upgrade : this can take a while.
+	for curStatus.LastVersion == initialStatus.LastVersion {
+		curStatus, err = client.Status()
+		require.NoError(t, err)
+
+		require.Less(t, int64(time.Now().Sub(startLoopTime)), int64(3*time.Minute))
+		time.Sleep(time.Duration(roundTimeMs) * time.Millisecond)
+		round = curStatus.LastRound
+	}
+
+	// now, that we have upgraded to the new protocol which supports rekey, try again.
+	_, err = client.BroadcastTransaction(rekey)
```

Comment:
I'm surprised you did not see any failures : The ordering of transaction within a block are not guaranteed.
You should wait until the first transaction is included in a block before sending the subsequent one.

Reply:
I guess they go into the same block since there no other traffic. I'll put them into a group

Reply:
ok, waiting is easier. Fixed.